### ISSUE TITLE
BUG: Application class resets API keys before every save.

### ISSUE THREAD

**Issue by anagrath** [strongloop/loopback#2096]
Here is code from git hub.  Before we go changing all the keys, shouldn't it check if the ctx.isNewInstance?  Difficult to inherit Application otherwise.

```
  /*!
   * A hook to generate keys before creation
   * @param next
   */
  Application.observe('before save', function(ctx, next) {
    if (!ctx.instance) {
      // Partial update - don't generate new keys
      // NOTE(bajtos) This also means that an atomic updateOrCreate
      // will not generate keys when a new record is creatd
      return next();
    }

    var app = ctx.instance;
    app.created = app.modified = new Date();
    if (!app.id) {
      app.id = generateKey('id', 'md5');
    }
    app.clientKey = generateKey('client');
    app.javaScriptKey = generateKey('javaScript');
    app.restApiKey = generateKey('restApi');
    app.windowsKey = generateKey('windows');
    app.masterKey = generateKey('master');
    next();
  });
```

**Comment by anagrath** [comment]
I have a work around:

`module.exports = (Model) ->
  Model.setup = () ->
    Model.app.models.Application.unobserve "before save"
    Model.app.models.Application.observe "before save", (ctx, next) ->
      if ctx.instance? && ctx.isNewInstance? && ctx.isNewInstance
        app = ctx.instance
        app.created = app.modified = new Date
        app.id = generateKey('id', 'md5') if !app.id
        app.clientKey = generateKey('client')
        app.javaScriptKey = generateKey('javaScript')
        app.restApiKey = generateKey('restApi')
        app.windowsKey = generateKey('windows')
        app.masterKey = generateKey('master')
      next()`

**Comment by loay** [comment]
Hi @anagrath I am not sure if I follow here. where is that code?
Is it a bug using a loopback app? Can you give more details so I can help out? thanks. 

Also, as for the before/after save hook, there is code in `common/models.user.js` about email update and it does trigger the newInstance. Thanks.

**Comment by anagrath** [comment]
Yes, sorry.  Now that I am looking at it again, I was also confused and I can see what happened.

The code I pasted was from Application within Loopback.  Every time you save Application, it generates new keys for oAuth.  Which means, if the user just changes the name of the application, too bad, all your old keys are invalid.  The work around I implemented to avoid this checks if there is a new instance.  So the question I was asking is, in the loopback code for the Application, shouldn't we only set new keys if it is a new application and then also only when we call for new keys via API?

The work around I implemented does that, but it seems loopback should behave that way since it is unexpected behavior for the Application to automatically take new keys.  I remember this was a difficult issue to debug because the oAuth keys for the devices I was configuring would be set, and then all of a sudden would no longer work for no explanation.
